File: packages/semantic-cache-py/betterdb_semantic_cache/adapters/anthropic.py
"""Anthropic Messages API adapter for betterdb-semantic-cache.

Extracts the text to embed from Anthropic Messages API request params.
Keys on the last user message's text content.

Usage::

    from betterdb_semantic_cache.adapters.anthropic import prepare_semantic_params
    params = await prepare_semantic_params(anthropic_params)
    result = await cache.check(params.text)
"""
from __future__ import annotations

import base64
from dataclasses import dataclass
from typing import Any

from ..normalizer import BinaryNormalizer, BinaryRef, default_normalizer
from ..utils import BinaryBlock, TextBlock


@dataclass
class SemanticParams:
    text: str
    blocks: list[TextBlock | BinaryBlock] | None = None
    model: str | None = None


def _get(obj: Any, key: str, default: Any = None) -> Any:
    if isinstance(obj, dict):
        return obj.get(key, default)
    return getattr(obj, key, default)


async def _normalize_block(
    block: Any,
    normalizer: BinaryNormalizer,
) -> TextBlock | BinaryBlock | None:
    block_type = _get(block, "type")

    if block_type == "text":
        return TextBlock(type="text", text=_get(block, "text", "") or "")

    if block_type == "image":
        src = _get(block, "source", {})
        src_type = _get(src, "type")
        media_type = "image/*"

        if src_type == "base64":
            source: Any = {"type": "base64", "data": _get(src, "data", "")}
            media_type = _get(src, "media_type") or "image/*"
        elif src_type == "url":
            source = {"type": "url", "url": _get(src, "url", "")}
        elif src_type == "file":
            source = {"type": "file_id", "file_id": _get(src, "file_id", ""), "provider": "anthropic"}
        else:
            return None

        ref = await normalizer({"kind": "image", "source": source})  # type: ignore[typeddict-item]
        return BinaryBlock(type="binary", kind="image", mediaType=media_type, ref=ref)

    if block_type == "document":
        src = _get(block, "source", {})
        src_type = _get(src, "type")
        media_type = "application/octet-stream"

        if src_type == "base64":
            source = {"type": "base64", "data": _get(src, "data", "")}
            media_type = _get(src, "media_type") or "application/pdf"
        elif src_type == "text":
            encoded = base64.b64encode((_get(src, "text") or "").encode()).decode()
            source = {"type": "base64", "data": encoded}
            media_type = "text/plain"
        elif src_type == "url":
            source = {"type": "url", "url": _get(src, "url", "")}
            media_type = "application/pdf"
        elif src_type == "file":
            source = {"type": "file_id", "file_id": _get(src, "file_id", ""), "provider": "anthropic"}
        else:
            return None

        ref = await normalizer({"kind": "document", "source": source})  # type: ignore[typeddict-item]
        return BinaryBlock(type="binary", kind="document", mediaType=media_type, ref=ref)

    return None


async def prepare_semantic_params(
    params: Any,
    *,
    normalizer: BinaryNormalizer | None = None,
) -> SemanticParams:
    """Extract semantic cache params from Anthropic Messages API request params."""
    norm = normalizer or default_normalizer
    messages = _get(params, "messages", [])
    model = _get(params, "model")

    user_messages = [m for m in (messages or []) if _get(m, "role") == "user"]
    if not user_messages:
        return SemanticParams(text="", model=model)

    last_user = user_messages[-1]
    content = _get(last_user, "content")

    if isinstance(content, str):
        return SemanticParams(text=content, model=model)

    if isinstance(content, list):
        blocks: list[TextBlock | BinaryBlock] = []
        for part in content:
            block = await _normalize_block(part, norm)
            if block is not None:
                blocks.append(block)
        text = " ".join(b["text"] for b in blocks if b.get("type") == "text")
        return SemanticParams(text=text, blocks=blocks, model=model)

    return SemanticParams(text="", model=model)

File: packages/semantic-cache/src/adapters/anthropic.ts
/**
 * Anthropic Messages API adapter for @betterdb/semantic-cache.
 *
 * Extracts the text to embed from Anthropic Messages API request params.
 * Semantic caching keys on the last user message's text content because that
 * is the actual query. See openai.ts for the full rationale.
 *
 * Usage:
 *   import { prepareSemanticParams } from '@betterdb/semantic-cache/anthropic';
 *   const { text, model } = await prepareSemanticParams(params);
 *   const result = await cache.check(text);
 */
import type {
  MessageCreateParamsNonStreaming,
  ContentBlockParam,
  TextBlockParam,
  ImageBlockParam,
  DocumentBlockParam,
} from '@anthropic-ai/sdk/resources';
import type { BinaryBlock, TextBlock } from '../utils';
import type { BinaryNormalizer, BinaryRef } from '../normalizer';
import { defaultNormalizer } from '../normalizer';

export interface AnthropicSemanticPrepareOptions {
  /** Binary content normalizer. Default: passthrough. */
  normalizer?: BinaryNormalizer;
}

export interface SemanticParams {
  text: string;
  blocks?: (TextBlock | BinaryBlock)[];
  model?: string;
}

async function normalizeBlock(
  block: ContentBlockParam,
  normalizer: BinaryNormalizer,
): Promise<TextBlock | BinaryBlock | null> {
  const type = (block as { type: string }).type;

  if (type === 'text') {
    const b = block as TextBlockParam;
    return { type: 'text', text: b.text };
  }

  if (type === 'image') {
    const b = block as ImageBlockParam;
    const src = b.source as { type: string; data?: string; media_type?: string; url?: string; file_id?: string };
    let source: BinaryRef['source'];
    let mediaType = 'image/*';

    if (src.type === 'base64') {
      source = { type: 'base64', data: src.data! };
      mediaType = src.media_type ?? 'image/*';
    } else if (src.type === 'url') {
      source = { type: 'url', url: src.url! };
    } else if (src.type === 'file') {
      source = { type: 'fileId', fileId: src.file_id!, provider: 'anthropic' };
    } else {
      return null;
    }

    const ref = await normalizer({ kind: 'image', source });
    return { type: 'binary', kind: 'image', mediaType, ref };
  }

  if (type === 'document') {
    const b = block as DocumentBlockParam;
    const src = b.source as {
      type: string;
      data?: string;
      text?: string;
      media_type?: string;
      url?: string;
      file_id?: string;
    };

    let source: BinaryRef['source'];
    let mediaType = 'application/octet-stream';

    if (src.type === 'base64') {
      source = { type: 'base64', data: src.data! };
      mediaType = src.media_type ?? 'application/pdf';
    } else if (src.type === 'text') {
      const encoded = Buffer.from(src.text!).toString('base64');
      source = { type: 'base64', data: encoded };
      mediaType = 'text/plain';
    } else if (src.type === 'url') {
      source = { type: 'url', url: src.url! };
      mediaType = 'application/pdf';
    } else if (src.type === 'file') {
      source = { type: 'fileId', fileId: src.file_id!, provider: 'anthropic' };
    } else {
      return null;
    }

    const ref = await normalizer({ kind: 'document', source });
    return { type: 'binary', kind: 'document', mediaType, ref };
  }

  return null;
}

/**
 * Extract semantic cache params from Anthropic Messages API request params.
 *
 * Extracts the last user message text for semantic similarity matching.
 * The system prompt is not included in the cache key because it rarely changes
 * within a deployment and would prevent hits across conversations.
 */
export async function prepareSemanticParams(
  params: MessageCreateParamsNonStreaming,
  opts?: AnthropicSemanticPrepareOptions,
): Promise<SemanticParams> {
  const normalizer = opts?.normalizer ?? defaultNormalizer;

  // Find last user message
  const userMessages = params.messages.filter((m) => m.role === 'user');
  if (userMessages.length === 0) {
    return { text: '', model: params.model };
  }

  const lastUser = userMessages[userMessages.length - 1];
  const content = lastUser.content;

  if (typeof content === 'string') {
    return { text: content, model: params.model };
  }

  if (Array.isArray(content)) {
    const blocks: (TextBlock | BinaryBlock)[] = [];
    for (const part of content as ContentBlockParam[]) {
      const block = await normalizeBlock(part, normalizer);
      if (block) blocks.push(block);
    }
    const text = blocks
      .filter((b): b is TextBlock => b.type === 'text')
      .map((b) => b.text)
      .join(' ');
    return { text, blocks, model: params.model };
  }

  return { text: '', model: params.model };
}
